core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,387 @@
+
+        return definitions.addAll(methodDefinitions);
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link MethodInfo}.
+     *
+     * @param methodInfo the method to generate the JSON Schema for.
+     *
+     * @return ObjectNode containing the JSON schema for the parameter type.
+     */
+    private ObjectNode generate(MethodInfo methodInfo) {
+        final ObjectNode root = mapper.createObjectNode();
+
+        root.put("$id", methodInfo.id())
+            .put("title", methodInfo.name())
+            .put("description", methodInfo.descriptionInfo().docString())
+            .put("additionalProperties", false)
+            // TODO: Assumes every method takes an object, which is only valid for RPC based services
+            //  and most of the REST services.
+            .put("type", "object");
+
+        final List<FieldInfo> methodFields;
+        final Map<TypeSignature, String> visited = new HashMap<>();
+        final String currentPath = "#";
+
+        if (methodInfo.useParameterAsRoot()) {
+            final TypeSignature signature = methodInfo.parameters().get(0)
+                                                      .typeSignature();
+            final StructInfo structInfo = typeSignatureToStructMapping.get(signature.signature());
+            if (structInfo == null) {
+                logger.info("Could not find root parameter with signature: {}", signature);
+                root.put("additionalProperties", true);
+                methodFields = Collections.emptyList();
+            } else {
+                methodFields = structInfo.fields();
+            }
+            visited.put(signature, currentPath);
+        } else {
+            methodFields = methodInfo.parameters();
+        }
+
+        generateProperties(methodFields, visited, currentPath, root);
+        return root;
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link FieldInfo} and add it to the given {@link ObjectNode}
+     * and add required fields to the {@link ArrayNode}.
+     *
+     * @param field field to generate schema for
+     * @param visited map of visited types and their paths
+     * @param path current path in tree traversal of fields
+     * @param parent the parent to add schema properties
+     * @param required the array node to add required field names
+     */
+    private void generateField(FieldInfo field, Map<TypeSignature, String> visited, String path,
+                               ObjectNode parent,
+                               ArrayNode required) {
+        final ObjectNode fieldNode = mapper.createObjectNode();
+        final TypeSignature fieldTypeSignature = field.typeSignature();
+
+        if (visited.containsKey(fieldTypeSignature)) {
+            // If field is already visited, add a reference to the field instead of iterating its children.
+            final String pathName = visited.get(fieldTypeSignature);
+            fieldNode.put("$ref", pathName);
```

Comment:
- Should `$ref` field be added to `required` if `field` is mandatory?
- `description` would be useful even for a reference field.